File: ios/Classes/VariableAppIconPlugin.m
#import "VariableAppIconPlugin.h"

@implementation VariableAppIconPlugin
+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
  FlutterMethodChannel* channel = [FlutterMethodChannel
      methodChannelWithName:@"variable_app_icon"
            binaryMessenger:[registrar messenger]];
  VariableAppIconPlugin* instance = [[VariableAppIconPlugin alloc] init];
  [registrar addMethodCallDelegate:instance channel:channel];
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
  if ([@"changeAppIcon" isEqualToString:call.method]) {
      NSString *iconName = (NSString*) call.arguments[@"iosIcon"];
      NSString *defaultiOSIcon = (NSString*) call.arguments[@"defaultiOS"];
      [UIApplication.sharedApplication setAlternateIconName:[iconName isEqualToString:defaultiOSIcon] ? nil : iconName completionHandler:nil];
  } else {
    result(FlutterMethodNotImplemented);
  }
}

@end

File: ios/Classes/VariableAppIconPlugin.swift
import Flutter
import UIKit

public class VariableAppIconPlugin: NSObject, FlutterPlugin {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "variable_app_icon", binaryMessenger: registrar.messenger())
    let instance = VariableAppIconPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {

    guard let arguments = call.arguments as? [String: Any] else { return }

    switch call.method {
    case "changeAppIcon":
      let defaultIcon = arguments["defaultiOS"] as! String
      let currentIcon = arguments["iosIcon"] as! String
      UIApplication.shared.setAlternateIconName(defaultIcon == currentIcon ? nil : currentIcon)

    default:
      result(FlutterMethodNotImplemented)
    }
  }
}
